In [2]:
import numpy as np
import netCDF4 as nc
import pandas as pd
import numpy.ma as ma

In [3]:
import matplotlib as mpl
import matplotlib.pyplot as plt

In [4]:
import xarray as xar

C:\ProgramData\Anaconda3\lib\site-packages\dask\config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  data = yaml.load(f.read()) or {}
C:\ProgramData\Anaconda3\lib\site-packages\distributed\config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  defaults = yaml.load(f)


In [5]:
file_path='./data/sanjioil.nc'

In [6]:
ds=nc.Dataset(file_path)

In [8]:
ds_xr=xar.open_dataset(file_path)

In [9]:
ds_xr

<xarray.Dataset>
Dimensions:                                   (time: 73, trajectory: 2958)
Coordinates:
  * trajectory                                (trajectory) int32 1 2 ... 2958
  * time                                      (time) datetime64[ns] 2018-01-14T22:20:00 ... 2018-01-17T22:20:00
    lon                                       (trajectory, time) float32 ...
    lat                                       (trajectory, time) float32 ...
Data variables:
    age_seconds                               (trajectory, time) float32 ...
    interfacial_area                          (trajectory, time) float32 ...
    mass_evaporated                           (trajectory, time) float32 ...
    water_fraction                            (trajectory, time) float32 ...
    y_wind                                    (trajectory, time) float32 ...
    sea_ice_area_fraction                     (trajectory, time) float32 ...
    density                                   (trajectory, time) float32 ...
    oil_film_thickness                        (trajectory, time) float32 ...
    x_sea_water_velocity                      (trajectory, time) float32 ...
    bulltime                                  (trajectory, time) float32 ...
    x_wind                                    (trajectory, time) float32 ...
    age_emulsion_seconds                      (trajectory, time) float32 ...
    sea_surface_wave_stokes_drift_y_velocity  (trajectory, time) float32 ...
    status                                    (trajectory, time) int32 ...
    viscosity                                 (trajectory, time) float32 ...
    sea_floor_depth_below_sea_level           (trajectory, time) float32 ...
    mass_dispersed                            (trajectory, time) float32 ...
    land_binary_mask                          (trajectory, time) float32 ...
    sea_surface_wave_significant_height       (trajectory, time) float32 ...
    sea_water_temperature                     (trajectory, time) float32 ...
    mass_oil                                  (trajectory, time) float32 ...
    age_exposure_seconds                      (trajectory, time) float32 ...
    sea_surface_wave_stokes_drift_x_velocity  (trajectory, time) float32 ...
    y_sea_water_velocity                      (trajectory, time) float32 ...
    z                                         (trajectory, time) float32 ...
    wind_drift_factor                         (trajectory, time) float32 ...
    fraction_evaporated                       (trajectory, time) float32 ...
Attributes:
    Conventions:                                  CF-1.6
    standard_name_vocabulary:                     CF-1.6
    featureType:                                  trajectory
    history:                                      Created 2019-08-01 11:42:45...
    source:                                       Output from simulation with...
    model_url:                                    https://github.com/OpenDrif...
    opendrift_class:                              OpenOil
    opendrift_module:                             opendrift.models.openoil
    readers:                                      ['E:\\project\\oilspill_res...
    time_coverage_start:                          2018-01-14 22:20:00
    time_step_calculation:                        0:30:00
    time_step_output:                             1:00:00
    config_seed:oil_type:                         MEDIUM (brage)
    config_seed:ocean_only:                       True
    config_general:use_basemap_landmask:          True
    config_general:basemap_resolution:            h
    config_general:coastline_action:              stranding
    config_general:time_step_minutes:             60
    config_general:time_step_output_minutes:      None
    config_drift:max_age_seconds:                 None
    config_drift:scheme:                          runge-kutta
    config_drift:stokes_drift:                    True
    config_drift:current_uncertainty:             0.5
    con

In [10]:
ds_xr.dims

Frozen(SortedKeysDict({'trajectory': 2958, 'time': 73}))

In [11]:
ds_xr.coords

Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
  * time        (time) datetime64[ns] 2018-01-14T22:20:00 ... 2018-01-17T22:20:00
    lon         (trajectory, time) float32 ...
    lat         (trajectory, time) float32 ...

In [12]:
ds_xr['mass_oil']

<xarray.DataArray 'mass_oil' (trajectory: 2958, time: 73)>
[215934 values with dtype=float32]
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
  * time        (time) datetime64[ns] 2018-01-14T22:20:00 ... 2018-01-17T22:20:00
    lon         (trajectory, time) float32 ...
    lat         (trajectory, time) float32 ...
Attributes:
    units:    kg

## 获取`mass_oil`的`DataArray`  
如下可知：
    `mass_oil`有2958个值，该物理量有两个维度，分别为`trajectory`,`time`

In [17]:
ds_xr['mass_oil']

<xarray.DataArray 'mass_oil' (trajectory: 2958, time: 73)>
[215934 values with dtype=float32]
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
  * time        (time) datetime64[ns] 2018-01-14T22:20:00 ... 2018-01-17T22:20:00
    lon         (trajectory, time) float32 ...
    lat         (trajectory, time) float32 ...
Attributes:
    units:    kg

### 1- 尝试将xarray转为series

In [11]:
ds_xr['mass_oil'].to_series()

trajectory  time               
1           2018-01-14 22:20:00    2.000000e+03
            2018-01-14 23:20:00    1.624608e+03
            2018-01-15 00:20:00    9.969210e+36
            2018-01-15 01:20:00    9.969210e+36
            2018-01-15 02:20:00    9.969210e+36
            2018-01-15 03:20:00    9.969210e+36
            2018-01-15 04:20:00    9.969210e+36
            2018-01-15 05:20:00    9.969210e+36
            2018-01-15 06:20:00    9.969210e+36
            2018-01-15 07:20:00    9.969210e+36
            2018-01-15 08:20:00    9.969210e+36
            2018-01-15 09:20:00    9.969210e+36
            2018-01-15 10:20:00    9.969210e+36
            2018-01-15 11:20:00    9.969210e+36
            2018-01-15 12:20:00    9.969210e+36
            2018-01-15 13:20:00    9.969210e+36
            2018-01-15 14:20:00    9.969210e+36
            2018-01-15 15:20:00    9.969210e+36
            2018-01-15 16:20:00    9.969210e+36
            2018-01-15 17:20:00    9.969210e+36
        

In [12]:
# 查看指定的val的长度
ds_xr.isel(time=0)['mass_oil'].sizes

Frozen({'trajectory': 2958})

## 2 剔除掉nan的数据，只获取有效数据

#### 查看不同时刻的
并通过`where`剔除掉大于`max`以及`nan`的值  
使用xarray的dropna方法，注意与pandas中的方法有所区别

提示错误：
`ValueError: any must be a single dataset dimension`  
查看文档后发现`xarray.Dataset.dropna`尚不支持同时沿多个维度放置  

[参考](http://xarray.pydata.org/en/stable/generated/xarray.Dataset.dropna.html)  

考虑将xarray转换为dateframe，然后再清洗？  
[上网参考](https://stackoverflow.com/questions/52553925/python-xarray-remove-coordinates-with-all-missing-variables)
比较好的解决办法是：
1. 将`DataSet`->`DataFrame`
2. 在`DataFrame`中通过`pandas` `dropna` 即可

In [23]:
xr_temp=ds_xr.isel(time=1)['mass_oil']
xr_temp

<xarray.DataArray 'mass_oil' (trajectory: 2958)>
array([1.624608e+03, 1.547655e+03, 1.677850e+03, ..., 9.969210e+36,
       9.969210e+36, 9.969210e+36], dtype=float32)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-14T23:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...
Attributes:
    units:    kg

In [34]:
#xr_temp.where(xr_temp<=2000).dropna(dim='any')

In [37]:
# 将xarray转成df
xr_temp.where(xr_temp<2000).to_dataframe().head()

time         lon        lat     mass_oil
trajectory                                                        
1          2018-01-14 23:20:00  125.912216  28.369555  1624.607544
2          2018-01-14 23:20:00  125.926674  28.373537  1547.655151
3          2018-01-14 23:20:00  125.915215  28.382929  1677.850220
4          2018-01-14 23:20:00  125.928246  28.359102  1556.745605
5          2018-01-14 23:20:00  125.900192  28.385672  1530.578369

* 这么看此种方式比较可行

In [44]:
xr_final=xr_temp.where(xr_temp<2000).to_dataframe().dropna(how='any')

In [48]:
xr_final.head()

time         lon        lat     mass_oil
trajectory                                                        
1          2018-01-14 23:20:00  125.912216  28.369555  1624.607544
2          2018-01-14 23:20:00  125.926674  28.373537  1547.655151
3          2018-01-14 23:20:00  125.915215  28.382929  1677.850220
4          2018-01-14 23:20:00  125.928246  28.359102  1556.745605
5          2018-01-14 23:20:00  125.900192  28.385672  1530.578369